### DISCUSSION TITLE
ENCHANT_CONFIG_DIR treatment

### DISCUSSION THREAD

**Discussion by shipmints** [minad/jinx#205] [Q&A]
I'd like to override ENCHANT_CONFIG_DIR in my Emacs config. I setenv this in early-init and I validate that this is set before package autoloads, certainly before jinx--load-module is called (I advised jinx--load-module to also illustrate that the env var is correctly set before its first call). I also validate that (fboundp #'jinx--mod-dict) is nil throughout the entire Emacs startup sequence until after emacs-startup-hook runs.

I do also observe that when the env var is set externally, it does take precedence, so I know libenchant honors the variable.

According to the libenchant source code, the value of the env var is dynamically applied at the point that a spelling provider is created.

I'm a bit puzzled since I expect libenchant to not be initalized until after jinx loads it and definitely by that time the env var is correctly set.

macos 12.7.5, emacs 29.3, latest jinx, latest enchant (via brew).

**Comment by minad** [comment]
You probably use `setenv` to alter the environment variable, which modifies the Emacs `process-environment`. But as far as I understand it doesn't modify the environment of the current Emacs process itself.

**Reply by shipmints** [reply]
Ach! Would you believe that I've been using Emacs for, um, quite the number of years, and this is the first time I've tried to setenv expecting it to be a simple wrapper around Posix setenv? I looked through the Emacs source code and there is no (easy) way to control the Posix environment of the Emacs process itself. Phooey.

In that case, I would like to humbly request that we expose this ability in jinx-mod.c so one can properly control the loaded library behavior without polluting their general environment? I'd be happy to submit a PR if you like.

I have no use for enchant-2/libenchant outside of Emacs. Since I share my Emacs config and packages with people on my team, I try my best to not impose additional changes on them, though installing libenchant to begin with is an imposition though one I automate away for them--but not their shell script set ups.

**Reply by minad** [reply]
To be honest, I consider such an addition out of scope, since the use case seems a bit too special. The addition works against the intended design of Enchant, where data is shared between applications. Overriding environment variables early on is easy enough, but I understand the implications in your case.

If my projects and help is useful for you and your team, I appreciate [support](https://github.com/sponsors/minad) regarding the development.

**Reply by shipmints** [reply]
Even the Enchant documentation says use symlinks if you want to share word files among packages so even those authors recognized that they can't be universal. For users on Macs that launch Emacs using the dock or Spotlight, environment variables are not easily controllable. This is the case here. Happy to submit a PR for this.

I'd also want to expose the libenchant version number to elisp so we can see it. And consider adding a suffix for the compiled module output name; e.g., (car (split-string system-configuration "-")) ; e.g., "x86_64" or "aarch64" to support sharing the Emacs directory hierarchy among architectures (which is what I do). The file would be called something like "jinx-mod.x86_64.dylib" or "jinx-mod-x86_64.dylib":
```elisp
    (let* ((mod-name (file-name-with-extension "jinx-mod" (concat (car (split-string system-configuration "-")) "-" module-file-suffix)))
```
This is how we segregate architecture-specific treesit grammars that share the same tree.
```elisp
  (add-to-list 'treesit-extra-load-path
    (expand-file-name (concat "treesit/" (car (split-string system-configuration "-"))) user-emacs-directory)))
```
P.S. I'll talk to the bosses about support.

**Reply by minad** [reply]
> Even the Enchant documentation says use symlinks if you want to share word files among packages so even those authors recognized that they can't be universal. For users on Macs that launch Emacs using the dock or Spotlight, environment variables are not easily controllable. This is the case here.

As far as I understand, symlinks should be used if you want to share wordlists among *different* packages, e.g., Enchant and some other spell checker. I am not fond of bypassing the Enchant API by setting an environment variable before initializing Enchant, but I'll think about it.

> I'd also want to expose the libenchant version number to elisp so we can see it. And consider adding a suffix for the compiled module output name; e.g., (car (split-string system-configuration "-")) ; e.g., "x86_64" or "aarch64" to support sharing the Emacs directory hierarchy among architectures (which is what I do). The file would be called something like "jinx-mod.x86_64.dylib" or "jinx-mod-x86_64.dylib":

Yes, we could do that. Note that you could also use different directories and then adjust the load-path accordingly.

In any case, the things you are asking for are specific to your setup/distribution model and specific to Mac. They will add to my overall maintenance effort, but I am happy to consider given some support.

> P.S. I'll talk to the bosses about support.

Thanks.

**Reply by shipmints** [reply]
These use-cases are not specific to Mac. They are specific to a) any platform where an app launcher does not start from one's .bashrc or .zshrc and includes Linux platform launchers, Microsoft Windows, Mac, etc. and in secure settings where system-level env vars aren't under user control b) anyone who uses multiple computers with differing platforms and wants to share the same underlying emacs/elpa tree and includes people who run Linux Intel and Linux ARM e.g., the popular Raspberry Pi platform--this is the same for .so .dll .dylib, etc. that all share the same directory and need differentiation. To support env var and the module file-name adornment I'm happy to provide PRs and support those features.

**Reply by shipmints** [reply]
Thanks, again, for jinx and all your other work. I finally made some time to offer a PR. See https://github.com/minad/jinx/pull/208. -Stephane

**Reply by minad** [reply]
Thanks. Please check out my [sponsoring](https://github.com/sponsors/minad) page. The professional tier seems appropriate in this case.

**Reply by shipmints** [reply]
I forgot to mention that I did check, and as this is in an academic environment, they don't have a budget for it. They seemed happy that I wrote the code and contributed it for inclusion. I have assigned FSF paperwork.

**Reply by shipmints** [reply]
Please accept these changes. I've been using them since the day I sent them in to you. Other people can benefit. And you, too, as it will lower support burden for people who share their emacs configs across machines (which is more people than you might think).

**Reply by shipmints** [reply]
FWIW, I just experimented with sponsoring on my own.

**Reply by minad** [reply]
Thanks. So to be clear, I am considering your contribution, otherwise we wouldn't have this conversation. However every contribution needs maintenance and curation. This means I have to rework your and similar contributions and bring them into a form which I find consider good for the project. This costs a lot of time. I am in a difficult personal situation right now - I don't want to go into details, but this leads to a scenario where I need more support and cannot give away my time for free. Now contrast this with the increase in user numbers of my packages, and the natural increase in feature requests which follows from that.

**Reply by shipmints** [reply]
Of course. I'm with you. I'm currently contributing to several open source projects that I also have to support (in fact, working on a tricky bug right this very second which is taking up more time than I'd prefer).

With specific regard to the jinx changes in question, I find them very small, easy to reason about, and an actual improvement to jinx considering the nature of dynamic module use. Please review the changes and accept them. If there's something you'd like me to change, I'm happy to do it. I'm also happy to help support these changes.

**Reply by shipmints** [reply]
Consider applying for a $$$ grant here https://floss.fund/blog/announcing-floss-fund/

**Reply by shipmints** [reply]
https://github.blog/news-insights/company-news/announcing-github-secure-open-source-fund/
